### DISCUSSION TITLE
DMA on ESP32-S3 is working

### DISCUSSION THREAD

**Discussion by stg** [Bodmer/TFT_eSPI#2233] [New processor/display support or capability request]
Not so much an issue as a potential solution.

I got DMA on ESP32-S3 working just fine, with some minor adjustments.
Since I am only using this library as a reference for writing a low-level implementation, I do not have a proper fix to share - but figured it might be worth something to you to know what I did to get it working.

1) Use HSPI
2) DMA_CHANNEL <= SPI_DMA_CH_AUTO
3) spi_host <= SPI3_HOST
4) After a DMA transfer: SPI_DMA_CONF_REG <= 0b11

I am doing step 4 in TFT_WRITE_BITS, which isn't a clean solution at all but works just fine, with HSPI at least.

**Comment by Bodmer** [comment]
Thanks for taking the time to post this. I will see if I can incorporate.

**Comment by stg** [comment]
No problem.

I would assume step 4 has to do with SPI_DMA_RX/TX_ENA and that my resetting the whole register just happened to also clear these bits.

The reason I think this is the case is the following from the ref. manual: `The bits SPI_DMA_RX_ENA and SPI_DMA_TX_ENA
in register SPI_DMA_CONF_REG should be set, otherwise the read/write data will be stored to/sent from
the registers SPI_W0_REG ~ SPI_W15_REG` 

When I had my logic analyzer connected I could see that all transactions were in fact taking place as expected - but there was no data (MISO line was kept low), which would be explained by the aforementioned statement.

Let me know if you need/want help with testing, since I have everything hooked up (for the next week or so).

**Reply by Bodmer** [reply]
Other RP2040 projects and library developments prevent me from working on this at the moment. Thanks for the offer of support but it will be some time before I can look at this in detail and progress it.

**Comment by Qvantos** [comment]
> Not so much an issue as a potential solution.
> 
> I got DMA on ESP32-S3 working just fine, with some minor adjustments. Since I am only using this library as a reference for writing a low-level implementation, I do not have a proper fix to share - but figured it might be worth something to you to know what I did to get it working.
> 
> 1. Use HSPI
> 2. DMA_CHANNEL <= SPI_DMA_CH_AUTO
> 3. spi_host <= SPI3_HOST
> 4. After a DMA transfer: SPI_DMA_CONF_REG <= 0b11
> 
> I am doing step 4 in TFT_WRITE_BITS, which isn't a clean solution at all but works just fine, with HSPI at least.

Please write where to enter these changes

**Reply by tgghtp** [reply]
Late reply, but maybe this workaround helps:

I had the same issue with an ESP32-S3-DevKitC-1_N16R8 clone: the first output works, but then it stops updating.
Setup:

    VSCode, PlatformIO, Framework = Arduino
    ST7796_DRIVER
    USE_HSPI_PORT
    MISO and MOSI connected to the display
    this display: https://www.amazon.de/dp/B0CZRV3Q8T?ref=ppx_yo2ov_dt_b_fed_asin_title

Library version:
 "name": "TFT_eSPI",
 "version": "2.5.43",

code fragment:

```
  tft.begin();
  tft.setRotation(3);
  tft.setTouch(myPrefs.calibrationData);

#ifdef USE_DMA
  tft.initDMA(); // To use SPI DMA you must call initDMA() to setup the DMA engine
#endif
  tft.fillScreen(TFT_WHITE);
  tft.setTextColor(TFT_BLACK, TFT_WHITE, true);
  tft.drawCentreString(SW_PRODUCTVERSION, tft.width() / 2, tft.height() / 2, 2);

==>>>From here on, no further updates to the display are visible

```
I changed this in 

**TFT_eSPI_ESP32_S3.h:**
```
#include "driver/spi_master.h"
#include "hal/gpio_ll.h"
add==>  #include "soc/soc.h"

```
**TFT_eSPI_ESP32_S3.c:**
```
void TFT_eSPI::dmaWait(void){
          ...
add=>>  WRITE_PERI_REG( SPI_DMA_CONF_REG(SPI_DMA_CH_AUTO) , 0x00 );
      spiBusyCheck = 0;
 }

```
I'm not sure if this is the best solution, but now the TFT works. (I didn’t measure drawing times to verify DMA, but it looks fast.)

**Reply by AlexBestoso** [reply]
The changes provided here confirmed functional for me (ESP32-S3-Wroom-1 || T-Deck)

Also was able to Confirm that the DMA initalization process used by the TFT_eSPI library works like a charm (Properly enables the GDMA registers in the esp32 system)

**Reply by tgghtp** [reply]
**Update:**
moved modification in TFT_eSPI_ESP32_S3.c to here: 

```
void IRAM_ATTR dma_end_callback(spi_transaction_t *spi_tx)
{
...
...
WRITE_PERI_REG( SPI_DMA_CONF_REG(SPI_DMA_CH_AUTO) , 0b11 );  //workaround
// WRITE_PERI_REG(SPI_DMA_CONF_REG(spi_host), 0);
}

```

**Tests:**

_Test code:_
```
  //test DMA with tft.fillscreen
  #define NUM_DMA_FILL_TESTS 50
    tft.fillScreen(TFT_WHITE);
    tft.drawCentreString("Test mit DMA", tft.width() / 2, tft.height() / 2, 2);
    delay(1000);
    unsigned long timestampFILL = millis();
    for (int i = 0; i < NUM_DMA_FILL_TESTS; i++)
      tft.fillScreen(random(0xFFFF));
    Serial.printf("mit tft.initDMA() NUM_DMA_FILL_TESTS:%d Test fillscreen speed: %d ms\n", NUM_DMA_FILL_TESTS, millis() - timestampFILL);

  //test DMA with PNG Images
  //output on tft stops here at 1st output of pushPixelsDMA on ESP32-S3 w/o modification in dma_end_callback
  //on ESP32 output ok
  #define NUM_DMA_IMAGE_TESTS_PNG 50
  unsigned long timestampIMAGE_PNG = millis();
  for(int i=0;i<50;i++)
  {
    setDABImage_PNG();
    setDABPlusImage_PNG();
    setFMImage_PNG();
  }
  Serial.printf("mit tft.initDMA() NUM_DMA_IMAGE_TESTS_PNG:%d Test image draw speed: %d ms\n", NUM_DMA_IMAGE_TESTS_PNG, millis() - timestampIMAGE_PNG);

  //test DMA with JPG Images
  #define NUM_DMA_IMAGE_TESTS_JPG 50
  unsigned long timestampIMAGE_JPG = millis();
  for(int i=0;i<50;i++)
  {
    setDABImage_JPG();
    setDABPlusImage_JPG();
    setFMImage_JPG();
  }
  Serial.printf("mit tft.initDMA() NUM_DMA_IMAGE_TESTS_JPG:%d Test image draw speed: %d ms\n", NUM_DMA_IMAGE_TESTS_JPG, millis() - timestampIMAGE_JPG);

```



**Results:**
**Test on ESP32-S3:**
w/o my modification:
```
mit tft.initDMA() NUM_DMA_FILL_TESTS:50 Test fillscreen speed: 2256 ms
success, output visible on tft
first call to image output with pushImageDMA/pushPixelsDMA stops further output on screen

PNG_handler (DMA) decode time 12 ms (no ouput on screen visible)
JPEG_handler (DMA) drawtime 579 ms (no ouput on screen visible)

mit tft.initDMA() NUM_DMA_IMAGE_TESTS_PNG:50 Test image draw speed: 8563 ms   (no ouput on screen visible)
mit tft.initDMA() NUM_DMA_IMAGE_TESTS_JPG:50 Test image draw speed: 84521 ms  (no ouput on screen visible)  
```
_**test with image output disabled w/o my modification:**_
success, output always visible on tft



**_with my modification:_**
```
mit tft.initDMA() NUM_DMA_FILL_TESTS:50 Test fillscreen speed: 2254 ms    
mit tft.initDMA() NUM_DMA_IMAGE_TESTS_PNG:50 Test image draw speed: 8337 ms   (ouput on screen visible)  
mit tft.initDMA() NUM_DMA_IMAGE_TESTS_JPG:50 Test image draw speed: 84169 ms  (ouput on screen visible)

PNG_handler (DMA) decode time 55 ms
JPEG_handler (DMA) drawtime 534 ms
```




**Same code running on ESP32-Wrover:** 
```
mit tft.initDMA() NUM_DMA_FILL_TESTS:50 Test fillscreen speed: 2171 ms
mit tft.initDMA() NUM_DMA_IMAGE_TESTS_PNG:50 Test image draw speed: 12610 ms
mit tft.initDMA() NUM_DMA_IMAGE_TESTS_JPG:50 Test image draw speed: 22062 ms   

PNG_handler (DMA) decode time 68 ms
JPEG_handler (DMA) drawtime 137 ms
```


Reason for stopping perhaps in pushImageDMA/pushPixelsDMA

**Reply by bluecraank** [reply]
I can confirm this works on ESP32-S3-Touch-LCD-1.28 from WaveShare!

**Reply by Basitadam** [reply]
Works on ESP32-S3-Zero board, too.  With ST77916 controller. Thank you @tgghtp

**Reply by woshimaniubi8** [reply]
Your modification worked! After the change, calling pushPixelsDMA() on my ESP32-S3 successfully displays the content.
你的修改起作用了，修改后在我的esp32-s3上调用pushPixelsDMA()就可以成功显示了

**Reply by electronicsguy** [reply]
Board:
Waveshare-ESP32-S3-AMOLED-1.91
Display: ST7735

While the code with the modifications works in DMA mode, the time obtained per loop() (using Arduino's millis()) remains pretty much the same. 

For the badgers gif animation example, I get 450ms without DMA. Actually with DMA, I get 453ms. So I don't know if the DMA is actually helping or not.

**Reply by deulis** [reply]
I can confirm, DMA is working on XIAO ESP32S3 with the modification on dma_end_callback function and the driver GC9A01 with the following configuration in the tft_setup.h file:
```
#define USER_SETUP_LOADED // <--- Just in case
#define GC9A01_DRIVER
#define USE_HSPI_PORT // <--- Must be here to works on ESP32S3

#define TFT_SCLK D8
#define TFT_MOSI D10
#define TFT_MISO -1  // (unused)
#define TFT_CS   -1  // (unused)
#define TFT_DC   D3      
#define TFT_RST  D1

#define SPI_FREQUENCY  40000000
#define SPI_READ_FREQUENCY  20000000
#define SPI_TOUCH_FREQUENCY  2500000

```

**Comment by Bodmer** [comment]
I have updated the latest Github master (2.5.2) to add DMA for the ESP32 S3 (@stg thanks for the pointers). This has been tested with the examples provided.

Note that this is for SPI displays only and DMA is NOT supported for the ESP32 S3 with parallel interface displays.

**Reply by MrSToneX** [reply]
i used TFT_eSPI@2.5.2 build ESP32-S3 target, but platformIO build log print:

TFT_eSPI_ESP32_S3.h:11:3: warning: #warning >>>>------>> DMA is not supported on the ESP32 S3 (possible future update) [-Wcpp]
  #warning >>>>------>> DMA is not supported on the ESP32 S3 (possible future update)

maybe it's not work?

**Reply by Bodmer** [reply]
DMA is supported now for SPI, I have corrected the warning.

**Reply by MrSToneX** [reply]
Thanks for your reply!
i used 1.28 inch round display with GC9A01 driver IC, used 4 line SPI connect to ESP32-S3 (MOSI SCLK DC CS)

**Reply by notmart** [reply]
with today's master, on an ESP32-S3 and GC9A01 as well, with all the DMA examples I'm getting either a white screen or random garbage on screen.

perhaps depends also on what pins are used?

At the moment using
#define TFT_MOSI 2 // In some display driver board, it might be written as "SDA" and so on.
#define TFT_SCLK 15
#define TFT_CS   17  // Chip select control pin
#define TFT_DC   16  // Data Command control pin
#define TFT_RST  4  // Reset pin (could connect to Arduino RESET pin)
#define TFT_BL   5  // LED back-light

which works fine without DMA

**Reply by Bodmer** [reply]
For some reason DMA requires that a MISO pin is defined. If you do not have a spare then you can allocate it to be the same as the TFT_CS line and it will work, e.g. in your case add the following to the setup file:
#define TFT_MISO 17

**Reply by notmart** [reply]
works perfectly, thanks :)

**Reply by andrewchilds** [reply]
Just noting here that not having `#define TFT_MISO 17` in my setup was the culprit in my case (ESP32-S3-MINI-1U + ST7789V2 240x280).

**Reply by rsiemens77** [reply]
I am on 2.5.33, and when I try to use DMA with HSPI and MISO, only the very first update goes to the display, further updates do nothing, although seem to be running.  It's the same others are reporting in this thread. Any ideas?

**Reply by bluecraank** [reply]
Same issue like @rsiemens77

**Comment by shb3014** [comment]
DMA using HSPI and setting MISO to the same pin as TFT_CS works just fine. But it seems that dma with FSPI / SPI2 doesn't work. Are there other settings required for FSPI to work?

**Reply by Bodmer** [reply]
Make sure[ this line](https://github.com/Bodmer/TFT_eSPI/blob/master/User_Setups/Setup70b_ESP32_S3_ILI9341.h#L30) is commented out and the[ correct pins](https://github.com/Bodmer/TFT_eSPI/blob/master/User_Setups/Setup70b_ESP32_S3_ILI9341.h#L8-L11) are used for the port.

**Reply by shb3014** [reply]
> Make sure[ this line](https://github.com/Bodmer/TFT_eSPI/blob/master/User_Setups/Setup70b_ESP32_S3_ILI9341.h#L30) is commented out and the[ correct pins](https://github.com/Bodmer/TFT_eSPI/blob/master/User_Setups/Setup70b_ESP32_S3_ILI9341.h#L8-L11) are used for the port.

With the provided settings, FSPI still doesn't work with dma. The root cause is that the MISO macro is replaced with MOSI when using FSPI if MISO is choosen as -1. Commenting out the related part in `TFT_eSPI_ESP32_S3.h` resolve the issue. Setting MISO to CS would not help. 
```diff
    #if defined(CONFIG_IDF_TARGET_ESP32S3) || defined(CONFIG_IDF_TARGET_ESP32S2)
      #if (TFT_MISO == -1)
-        #undef TFT_MISO
-        #define TFT_MISO TFT_MOSI
+        //#undef TFT_MISO
+        //#define TFT_MISO TFT_MOSI
      #endif
    #endif
```

This seems to be related to the MISO init process in `spi_common.c`, specifically the `esp_rom_gpio_connect_out_signal` function. But this somehow does not affect HSPI.
```c++
if (bus_config->miso_io_num >= 0) {
            if (miso_need_output || (temp_flag&SPICOMMON_BUSFLAG_DUAL)) {
                gpio_set_direction(bus_config->miso_io_num, GPIO_MODE_INPUT_OUTPUT);
                esp_rom_gpio_connect_out_signal(bus_config->miso_io_num, spi_periph_signal[host].spiq_out, false, false);
            } else {
                gpio_set_direction(bus_config->miso_io_num, GPIO_MODE_INPUT);
            }
            esp_rom_gpio_connect_in_signal(bus_config->miso_io_num, spi_periph_signal[host].spiq_in, false);
#if CONFIG_IDF_TARGET_ESP32S2
            PIN_INPUT_ENABLE(GPIO_PIN_MUX_REG[bus_config->miso_io_num]);
#endif
            gpio_hal_iomux_func_sel(GPIO_PIN_MUX_REG[bus_config->miso_io_num], FUNC_GPIO);
        }
```
Also, it seems that setting MISO to the same as CS is not necessary with HSPI. Setting it to -1 would do the job.

**Reply by Bodmer** [reply]
Which Arduino ESP32 board package version are you using?  There have been some breaking changes recently!

**Reply by shb3014** [reply]
I use an esp32s3 module in a customly designed board. I'm also at the latest master branch.

**Comment by JackIsGoofingOff** [comment]
If you are using a round display from WaveShare, here is a quick solution (Add a new .h file in User_Setups and fill these codes and include this file in User_Setup_Select.h).

#define USER_SETUP_ID 305
#define USER_SETUP_INFO "ESP32-S3-LCD-1.28"

#ifdef ILI9341_DRIVER
#undef ILI9341_DRIVER
#endif

#define GC9A01_DRIVER

#define TFT_WIDTH  240
#define TFT_HEIGHT 240

#define TFT_MOSI 11
#define TFT_SCLK 10
#define TFT_CS   9
#define TFT_DC   8
#define TFT_RST  12
#define TFT_BL   40
#define TFT_BACKLIGHT_ON HIGH

#define LOAD_GLCD
#define LOAD_FONT2
#define LOAD_FONT4
#define LOAD_FONT6 
#define LOAD_FONT7
#define LOAD_FONT8
#define LOAD_FONT8N
#define LOAD_GFXFF 

#define SMOOTH_FONT

#define SPI_FREQUENCY  80000000

// related to DMA (direct memory allocation)
#define USE_HSPI_PORT
#define TFT_MISO  9

**Reply by JackIsGoofingOff** [reply]
But only image can be shown, animation is not working

**Reply by mjocean** [reply]
I'm having a problem which I think is similar to the last poster.

I am using a Waveshare ESP32-S3-Zero (the zero has no USB to UART chip and fewer exposed GPIO pins) and a separate GC9A01 1.28" round LCD display.  

Without DMA all is working after I added USE_HSPI_PORT and to knock my frequency down to 27MHz to my config.  Pins are as follows:
```
#define TFT_MOSI 11 //  "SDA" 
#define TFT_SCLK 10
#define TFT_CS    9  // Chip select control pin
#define TFT_DC    8  // Data Command control pin
#define TFT_RST   12  // Reset pin  --I have no header for pin 14 on this board
#define TFT_BL 40
#define TFT_MISO 9 // same as CS
```

Enabling DMA results in no updates to the display, though the program continues to run and logging shows that time is taken per pushImageDMA as there is a frame rate shown.  The same is true in my own code, where I see that pushPixelsDMA takes time, but on my code the display is still blank (same code works with our without DMA on the RP2040, with a different Setup, naturally).

To minimize variables: when running the DMA cube example source, a single initial yellow rect is shown on the display and it is frozen --it never moves, but on Serial reasonable FPS values are shown.  These values are slightly higher than those shown with DMA disabled, but with DMA disabled, the cube renders properly, moving around the screen.

Does anyone have any suggestions or advice?  Thank you!

**Comment by rsiemens77** [comment]
I am having the same issue on S3 as the previous recent reports, even with HSPI and MISO defined.  The first DMA update works, but successive updates do nothing, although all seems to be executing.  This exact DMA code works fine on ESP32E.  If I switch to non-DMA pushSprite() everything works fine as well.  

This sounds the same as the previous reports that also suggest first display works, but all further updates fail:

"But only image can be shown, animation is not working".

"when running the DMA cube example source, a single initial yellow rect is shown on the display and it is frozen --it never moves"

**Reply by rsiemens77** [reply]
I did a test where I mix some non-DMA updates with DMA updates.  The non-DMA updates work, until a DMA update is made.  That DMA update gets pushed to the display, but then -ALL- further updates stop, as if the LCD is frozen.  But the code seems to keep running fine.

**Reply by rsiemens77** [reply]
I put some logging messages in pushImageDMA and spi_device_queue_trans to see if I could find anything odd.

Both the first push and successive ones look the same, all the expected code seems to run, just no update after the first one.  And the LCD is frozen now, as non-DMA updates don't work either.

> I (9545) TFT: pushImageDMA
I (9548) TFT: dmaWait
I (9551) TFT: dmaWaitdone
I (9554) TFT: spiBusyChecked
I (9558) TFT: while
I (9562) TFT: while done
I (9563) TFT: queue_trans
I (9567) SPI: queue_trans
I (9570) SPI: bus acquired
I (9573) SPI: Queue Send
I (9577) SPI: lock bg request

**Comment by Basitadam** [comment]
After the first DMA transfer, even this does not recover the SPI communication:

tft.deInitDMA();
delay(100);
tft.begin();

(in fact, CS, DC and SCLK works as it should, but not MOSI. Stays at low level) - My board is ESP32-S3-Zero

**Comment by PhonicUK** [comment]
I'm having similar issues. Using `ST7796_DRIVER`  on an ESP32S3

```c
#define TFT_CS   10   // Chip Select control pin
#define TFT_RST  9   // Reset pin
#define TFT_DC   14   // Data/Command control pin (LCD_RS)
#define TFT_MOSI 11   // Master Out Slave In (SPI data line / SDI)
#define TFT_SCLK 12   // Clock pin for SPI
#define TFT_MISO 13    // Master In Slave Out (optional)
...
#define USE_HSPI_PORT
```

One transfer goes through, and then the whole thing immediately dies.

**Comment by dcodeIO** [comment]
Not sure if it helps, but after DMA was working fine, by just updating toolkits but no code, it stopped working for me, now exhibiting the same issue as described above in that the display freezes after the first DMA operation. Makes me wonder if the issue is framework version related.

**Comment by Elizafox** [comment]
I tried the JPEG DMA example with ESP32-C3. I set MOSI to CS as instructed above.

I am using Arduino IDE and an otherwise unmodified TFT_eSPI and an ST7789V with `ST7789_DRIVER`.

My pins are laid out:

```C++
#define TFT_MOSI 11
#define TFT_SCLK 10
#define TFT_CS 9
#define TFT_DC 8
#define TFT_RST 12
#define TFT_BL 40
#define TFT_BACKLIGHT_ON HIGH
```

I notice that 2.0.14 is the last working Arduino ESP32 board package version. Newer versions fail. I don't know what changed between 2.0.14 and 2.0.15 to cause this. I will investigate as I have time. This may be a good starting point: https://github.com/espressif/arduino-esp32/compare/2.0.14...2.0.15

**Comment by woshimaniubi8** [comment]
确实，虽然成功运行，但是貌似并没有对我的程序起帮助，这在我的LVGL程序上很容易看出来:帧率没有任何提升
True, it runs successfully, but it doesn't seem to have helped improve my program at all. This is clearly evident in my LVGL program: there's been no frame rate increase.
发自我的小米在 Sujay Phadke ***@***.***>，2025年6月21日 下午9:05写道：
Board:
Waveshare-ESP32-S3-AMOLED-1.91
Display: ST7735
While the code with the modifications works in DMA mode, the time obtained per loop() (using Arduino's millis()) remains pretty much the same.
For the badgers gif animation example, I get 450ms without DMA. Actually with DMA, I get 453ms. So I don't know if the DMA is actually helping or not.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you commented.Message ID: ***@***.***>
[
{
***@***.***": "http://schema.org",
***@***.***": "EmailMessage",
"potentialAction": {
***@***.***": "ViewAction",
"target": "https://github.com/Bodmer/TFT_eSPI/discussions/2233#discussioncomment-13537696",
"url": "https://github.com/Bodmer/TFT_eSPI/discussions/2233#discussioncomment-13537696",
"name": "View Discussion"
},
"description": "View this Discussion on GitHub",
"publisher": {
***@***.***": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

**Comment by electronicsguy** [comment]
Board:
Waveshare-ESP32-S3-AMOLED- 1.91 Display:
ST7735
While the code with the modifications works in DMA mode, the time obtained per loop() (using Arduino's millis()) remains pretty much the same.
For the badgers gif animation example, I get 450ms without DMA. Actually with DMA, I get 453ms. So I don't know if the DMA is actually helping or not.

I tried the built-in TFT_eSPI example, FLASH_JPG, with similar results:
DMA: 72 ms
without DMA: 67 ms

@stg @Bodmer

**Reply by stg** [reply]
It's not about the time it takes to transfer a frame; it's about the fact that the processor is free to do other things while that frame is transferring. Without DMA, the processor has to "manually" do the transfer, word for word, which basically means it's stuck in a data transfer loop { wait for SPI FIFO free; send word; } while (more data remaining). With DMA, the transfer happens in the "background," and the processor can do other things - such as rendering the *next* frame so that it's ready for DMA transfer as soon as the previous frame finishes. Normally, you double-buffer, such that you transfer one buffer while drawing to another and flip them after each frame. This makes a huge difference.

A decent FIFO provides a middle-ground; an interrupt must run to fill the FIFO every FIFO_SIZE frames minimum, where the CPU is still responsible for moving the data. That allows the processor to do other things, but it's not as "hands-off" as DMA is and will have a non-insignificant effect on available CPU time. It is also more susceptible to under-runs, which can affect timing and transfer speed unless the system is carefully designed. DMA can also have an effect on CPU time, due to memory contention, but it's usually invisible or very close to it. If the system supports DMA, it is almost always a better option.

The actual transfer time is primarily tied to frame size and SPI rate. Setting up DMA is a little extra work and will take a little extra time. This library uses Arduino and ESP-IDF hardware abstraction layers, which are a bit slow since they are designed for portability, thread safety, and resource sharing. An application-specific, tightly configured (no libs), DMA double-buffer system will send frames back-to-back with not as much as a clock cycle wasted between the transfers, achieving maximum theoretical throughput while only using a fraction of the available CPU time.

**Comment by Dakkaron** [comment]
According to https://docs.espressif.com/projects/esp-idf/en/v5.1.7/esp32s3/api-reference/peripherals/lcd.html, the ESP32-S3 should have a few more options for DMA than just SPI. I80 and 8/16-bit-parallel RGB should work too.

**Reply by Bodmer** [reply]
Yes, I have a QSPI display but have not got around to integrating a driver. I tried and failed to get I80 runiing when the feature was introduced, but I see it looks like debugged example code is availble now available. I was mostly trying to work with a datasheet lacking some important details!

**Reply by Dakkaron** [reply]
I have a Lilygo T-HMI with 8-bit parallel non-I80. It's on my todo-list to implement DMA with TFT_eSPI, but I don't know when I get around to it. But at least the example doesn't look difficult at all by now.
